File: src/lib/ipc/artifacts.ts
import { invoke } from '@tauri-apps/api/core';

/**
 * Mirrors `src-tauri/src/artifacts/mod.rs::ArtifactInfo`. Sent
 * snake_case on the wire because the Rust `serde::Serialize`
 * derive doesn't apply a rename_all attribute — keep them aligned
 * if either side ever switches.
 */
export interface ArtifactInfo {
  run_id: string;
  name: string;
  /** Absolute path on disk; ok to surface in the GUI for "reveal"
   *  affordances, never sent to remote services. */
  path: string;
  /** Byte count. */
  size: number;
  /** Last-modified unix ms. */
  mtime_ms: number;
}

/** List artifacts for one run. Empty array when the run has no
 *  artifacts yet (the dir doesn't exist) — not an error. */
export function artifactList(runId: string): Promise<ArtifactInfo[]> {
  return invoke<ArtifactInfo[]>('artifact_list', { runId });
}

/** Resolve absolute path to one named artifact under `runId`. */
export function artifactPath(runId: string, name: string): Promise<string> {
  return invoke<string>('artifact_path', { runId, name });
}

/** Write text content as an artifact. Use for power-user flows that
 *  write structured outputs without going through Hermes Agent. */
export function artifactWrite(
  runId: string,
  name: string,
  content: string,
): Promise<ArtifactInfo> {
  return invoke<ArtifactInfo>('artifact_write', { runId, name, content });
}

File: src-tauri/src/ipc/artifacts.rs
//! IPC for B-9.4 workflow artifacts. Three commands cover the full
//! GUI surface:
//!
//! - `artifact_list(run_id)` — enumerate files under
//!   `~/.hermes/artifacts/<run_id>/` with size + mtime
//! - `artifact_path(run_id, name)` — resolve an absolute path so the
//!   shell can open / reveal the file
//! - `artifact_write(run_id, name, content)` — write text content;
//!   exposed for power-user workflow YAMLs that want to publish
//!   structured outputs without hopping through MCP

use crate::artifacts::{self, ArtifactInfo};
use crate::error::{IpcError, IpcResult};

#[tauri::command]
pub async fn artifact_list(run_id: String) -> IpcResult<Vec<ArtifactInfo>> {
    tokio::task::spawn_blocking(move || artifacts::list_artifacts(&run_id))
        .await
        .map_err(|e| IpcError::Internal {
            message: format!("artifact_list join: {e}"),
        })?
        .map_err(|e| IpcError::Internal {
            message: format!("list_artifacts: {e}"),
        })
}

#[tauri::command]
pub async fn artifact_path(run_id: String, name: String) -> IpcResult<String> {
    tokio::task::spawn_blocking(move || artifacts::artifact_path(&run_id, &name))
        .await
        .map_err(|e| IpcError::Internal {
            message: format!("artifact_path join: {e}"),
        })?
        .map(|p| p.to_string_lossy().into_owned())
        .map_err(|e| IpcError::Internal {
            message: format!("artifact_path: {e}"),
        })
}

#[tauri::command]
pub async fn artifact_write(
    run_id: String,
    name: String,
    content: String,
) -> IpcResult<ArtifactInfo> {
    tokio::task::spawn_blocking(move || {
        artifacts::write_artifact(&run_id, &name, content.as_bytes())
    })
    .await
    .map_err(|e| IpcError::Internal {
        message: format!("artifact_write join: {e}"),
    })?
    .map_err(|e| IpcError::Internal {
        message: format!("write_artifact: {e}"),
    })
}
